File: backend/internal/signing/signing.go
// Package signing is the org-scoped qualified-document-signing slice. The
// business wallet is the RP-centric Signature Creation Application (SCA): it
// prepares a PDF, drives the QTSP's OAuth2/OID4VP authorization (the wallet
// ceremony), asks the QTSP to sign the document hash, and assembles the signed
// PAdES itself — the document never leaves the wallet.
//
// Layering (handler -> service -> store / provider): the provider seam is
// internal/signingprovider (a CSC/OAuth client or its stub); the per-org
// connection (base URL, client id, client secret) comes from internal/csc.
//
// Scope: PAdES-B/B-T (a detached CMS signature over the ByteRange). B-LT/B-LTA
// (validation data, archival timestamp) are out of scope and belong to a
// separate DSS augmentation seam.
package signing

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"

	"github.com/privacybydesign/yivi-businesswallet/backend/internal/signingprovider"
)

// DefaultRedirectURI is the QTSP-registered OAuth callback the browser is sent
// back to after the wallet ceremony. It must match a redirect-uri in the
// authorization server's client registration (see the --profile signer demo's
// config/application-secret.yml). This dev default targets the local backend;
// a real deployment behind a different host would need this made configurable.
const DefaultRedirectURI = "http://localhost:8080/api/v1/signing/callback"

// SessionTTL bounds how long a signing/link authorization may stay in flight
// before the in-memory session is abandoned and the request fails. It is the
// window a person has to complete the wallet ceremony.
const SessionTTL = 5 * time.Minute

// Request statuses. A request is awaiting_signatures until every selected signer
// has signed; then completed (delivery, if any, is tracked separately on
// DeliveryStatus). failed means a signer's ceremony errored out.
const (
	StatusAwaitingSignatures = "awaiting_signatures"
	StatusCompleted          = "completed"
	StatusFailed             = "failed"
)

// Signing modes: parallel signers may sign in any order (serialized by an
// in-flight lock, since incremental PAdES cannot merge concurrent signatures);
// sequential signers must sign in ascending sign_order.
const (
	ModeParallel   = "parallel"
	ModeSequential = "sequential"
)

// Recipient channels the finished document may be delivered over.
const (
	ChannelNone  = "none"
	ChannelQERDS = "qerds"
	ChannelEmail = "email"
)

// Per-request delivery lifecycle (independent of the signing status).
const (
	DeliveryNotRequested = "not_requested"
	DeliveryPending      = "pending"
	DeliveryDelivered    = "delivered"
	DeliveryFailed       = "failed"
)

// Per-signer statuses.
const (
	SignerPending = "pending"
	SignerSigned  = "signed"
	SignerFailed  = "failed"
)

var (
	// ErrNotConfigured means the org has no usable CSC provider configured.
	ErrNotConfigured = errors.New("signing: no CSC signing provider configured for organization")
	// ErrNoCredential means the acting user has not linked a signing credential.
	ErrNoCredential = errors.New("signing: no signing credential linked; link one first")
	// ErrNotFound means a signing request does not exist (or is not the caller's).
	ErrNotFound = errors.New("signing: signing request not found")
	// ErrNotCompleted means the signed document is not ready yet.
	ErrNotCompleted = errors.New("signing: signing request is not completed")
	// ErrSessionExpired means the authorization was not completed within SessionTTL.
	ErrSessionExpired = errors.New("signing: authorization session expired")
	// ErrInvalidPDF means the uploaded bytes are not a usable PDF.
	ErrInvalidPDF = errors.New("signing: the uploaded file is not a valid PDF")
	// ErrNotSigner means the acting user is not a signer of the request.
	ErrNotSigner = errors.New("signing: you are not a signer of this request")
	// ErrAlreadySigned means the acting user has already signed the request.
	ErrAlreadySigned = errors.New("signing: you have already signed this request")
	// ErrNotYourTurn means an earlier signer must sign first (sequential mode).
	ErrNotYourTurn = errors.New("signing: it is not your turn to sign yet")
	// ErrSignInProgress means another signer's ceremony is in flight for this request.
	ErrSignInProgress = errors.New("signing: another signature is in progress for this document")
	// ErrInvalidRequest means the request parameters (signers, mode, recipient) are invalid.
	ErrInvalidRequest = errors.New("signing: invalid signing request")
)

// provider is the consumer-defined view of internal/signingprovider that this
// package depends on (accept interfaces, return structs). Both the concrete
// Client and the StubProvider satisfy it.
type provider interface {
	Discover(ctx context.Context, baseURL string) (signingprovider.Info, error)
	AuthorizeURL(issuer string, p signingprovider.AuthorizeParams) string
	ExchangeToken(ctx context.Context, issuer, clientID, clientSecret, code, codeVerifier, redirectURI string) (signingprovider.Token, error)
	ListCredentials(ctx context.Context, baseURL, accessToken string) ([]string, error)
	CredentialInfo(ctx context.Context, baseURL, accessToken, credentialID string) (signingprovider.Credential, error)
	SignHash(ctx context.Context, baseURL, accessToken, credentialID string, hashesB64 []string, signAlgoOID, hashAlgoOID string) ([]string, error)
}

// connectionResolver resolves an org's CSC connection (base URL + OAuth client
// credentials). internal/csc.Store implements it; the client secret is decrypted
// inside that store and only handed here for the in-flight token exchange.
type connectionResolver interface {
	ResolveConnection(ctx context.Context, orgID uuid.UUID) (baseURL, clientID, clientSecret string, err error)
	// Available reports whether the org has a signing provider configured and
	// enabled. It is member-safe (no secret), so the signing feature can be gated
	// for members who cannot read the admin-only provider settings.
	Available(ctx context.Context, orgID uuid.UUID) (bool, error)
}

// OrgMember is the minimal view of an org member this package needs to validate
// selected signers and label them in the UI. The adapter in cmd/api maps
// organization.Member onto it, keeping this slice free of that import.
type OrgMember struct {
	UserID uuid.UUID
	Name   string
	Email  string
}

// memberDirectory lists an org's active members, so a create-request call can
// validate the chosen signer ids and enrich signers with a name/email for display.
type memberDirectory interface {
	ListMembers(ctx context.Context, orgID uuid.UUID) ([]OrgMember, error)
}

// documentDeliverer delivers the finished signed PDF to an external natural
// person over the recipient's chosen channel. The adapters in cmd/api wrap
// email.Service (attachment) and qerds.Service (registered delivery). Delivery is
// best-effort: a failure is recorded on the request (delivery_status=failed), not
// fatal to the completed signature.
type documentDeliverer interface {
	DeliverEmail(ctx context.Context, orgID uuid.UUID, to, recipientName, coverMessage, filename string, pdf []byte) error
	DeliverQERDS(ctx context.Context, orgID uuid.UUID, to, recipientName, subject, coverMessage, filename string, pdf []byte) error
}

// signerNotifier tells a selected member that a document is waiting for their
// signature. The adapter in cmd/api wraps email.Service and builds the signing-page
// link from the org slug. Notification is best-effort: a failure is logged, never
// fatal to creating (or advancing) the request.
type signerNotifier interface {
	NotifySignatureRequested(ctx context.Context, orgID uuid.UUID, signerEmail, documentName, slug string) error
}

// LinkedCredential is a user's cached signing credential (fetched once, so the
// signing ceremony is a single authorize — the cert must be known before the
// document hash is computed, because the CMS SignedAttributes hash the cert).
type LinkedCredential struct {
	CredentialID string    `json:"credentialId"`
	KeyAlgo      string    `json:"keyAlgo"`
	UpdatedAt    time.Time `json:"updatedAt"`
}

// Signer is one member who must sign a request, with their per-signer state.
type Signer struct {
	UserID   uuid.UUID  `json:"userId"`
	Name     string     `json:"name"`
	Email    string     `json:"email"`
	Order    int        `json:"order"`
	Status   string     `json:"status"`
	SignedAt *time.Time `json:"signedAt,omitempty"`
}

// SignerInput is one selected signer at create time (order is 1-based; ignored in
// parallel mode).
type SignerInput struct {
	UserID uuid.UUID
	Order  int
}

// RecipientInput is where the finished document is delivered, chosen at create time.
type RecipientInput struct {
	Channel string // ChannelNone | ChannelQERDS | ChannelEmail
	Address string // email address or QERDS address; empty for ChannelNone
	Name    string
	Message string // cover message shown to the recipient
}

// Request is the non-document view of a signing request.
type Request struct {
	ID               uuid.UUID  `json:"id"`
	Status           string     `json:"status"`
	Filename         string     `json:"filename"`
	Mode             string     `json:"mode"`
	CreatedBy        uuid.UUID  `json:"createdBy"`
	CreatedByName    string     `json:"createdByName"`
	RecipientChannel string     `json:"recipientChannel"`
	RecipientName    string     `json:"recipientName,omitempty"`
	RecipientAddress string     `json:"recipientAddress,omitempty"`
	Message          string     `json:"message,omitempty"`
	DeliveryStatus   string     `json:"deliveryStatus"`
	DeliveryError    string     `json:"deliveryError,omitempty"`
	Error            string     `json:"error,omitempty"`
	Signers          []Signer   `json:"signers"`
	CreatedAt        time.Time  `json:"createdAt"`
	CompletedAt      *time.Time `json:"completedAt,omitempty"`
}

// Start is what StartLink/StartSign return to the frontend: the authorization URL
// to send the browser to, plus (for signing) the request id being signed.
type Start struct {
	RequestID    *uuid.UUID `json:"requestId,omitempty"`
	AuthorizeURL string     `json:"authorizeUrl"`
}

File: frontend/src/api/signing.ts
import { z } from "zod";
import { ApiError, absoluteApiUrl, request, requestBlob } from "./http";

// The qualified co-signing workflow: the business wallet is the RP-centric
// Signature Creation Application. Each signer links a signing credential once (a
// member in the "My credential" tab, an external signee behind their invitation
// link), then a request routes a PDF to its signers; each signs via an OID4VP + CSC
// ceremony that hands the browser off to the QTSP authorization server and returns
// here. When all have signed, the document is delivered to the recipient over QERDS
// or email.

const NOT_FOUND_STATUS = 404;

export const linkedCredentialSchema = z.object({
  credentialId: z.string(),
  keyAlgo: z.string(),
  updatedAt: z.string(),
});
export type LinkedCredential = z.infer<typeof linkedCredentialSchema>;

// Request statuses mirror the backend (internal/signing).
export const SIGNING_STATUS = {
  awaitingSignatures: "awaiting_signatures",
  completed: "completed",
  failed: "failed",
} as const;

// Signing modes, recipient channels, per-signer and delivery statuses — all mirror
// the backend. Kept as plain string maps (not zod enums) so a new backend value
// never breaks response parsing.
export const SIGNING_MODE = {
  parallel: "parallel",
  sequential: "sequential",
} as const;
export type SigningMode = (typeof SIGNING_MODE)[keyof typeof SIGNING_MODE];

export const RECIPIENT_CHANNEL = {
  none: "none",
  qerds: "qerds",
  email: "email",
} as const;
export type RecipientChannel =
  (typeof RECIPIENT_CHANNEL)[keyof typeof RECIPIENT_CHANNEL];

export const SIGNER_STATUS = {
  pending: "pending",
  signed: "signed",
  failed: "failed",
} as const;

export const DELIVERY_STATUS = {
  notRequested: "not_requested",
  pending: "pending",
  delivered: "delivered",
  failed: "failed",
} as const;

// A signer is either an internal org member or an external signee; only the first
// has a userId, so the signer row's own id is what identifies either.
export const SIGNER_KIND = {
  internal: "internal",
  external: "external",
} as const;
export type SignerKind = (typeof SIGNER_KIND)[keyof typeof SIGNER_KIND];

export const signerSchema = z.object({
  id: z.string(),
  kind: z.string(),
  userId: z.string().optional(),
  name: z.string(),
  email: z.string(),
  order: z.number(),
  status: z.string(),
  signedAt: z.string().optional(),
});
export type Signer = z.infer<typeof signerSchema>;

export const signingRequestSchema = z.object({
  id: z.string(),
  status: z.string(),
  filename: z.string(),
  mode: z.string(),
  createdBy: z.string(),
  createdByName: z.string(),
  recipientChannel: z.string(),
  recipientName: z.string().optional(),
  recipientAddress: z.string().optional(),
  message: z.string().optional(),
  deliveryStatus: z.string(),
  deliveryError: z.string().optional(),
  error: z.string().optional(),
  signers: z.array(signerSchema).nullable().default([]),
  createdAt: z.string(),
  completedAt: z.string().optional(),
});
export type SigningRequest = z.infer<typeof signingRequestSchema>;

const signingRequestListSchema = z.object({
  requests: z.array(signingRequestSchema).nullable().default([]),
});

const signingRequestPageSchema = z.object({
  requests: z.array(signingRequestSchema).nullable().default([]),
  nextCursor: z.string(),
});
export type SigningRequestPage = z.infer<typeof signingRequestPageSchema>;

// startLink / startSign both return the URL to send the browser to; signing also
// returns the request id being signed so the page can track it on return.
export const signingStartSchema = z.object({
  requestId: z.string().optional(),
  authorizeUrl: z.string(),
});
export type SigningStart = z.infer<typeof signingStartSchema>;

const createdSchema = z.object({ id: z.string() });

// SignerSelection is one chosen signer on the create form. Members and external
// signees share one ordered list, because that order is the sequential signing order
// and two separate lists could not express an order across both.
export type SignerSelection =
  | { kind: typeof SIGNER_KIND.internal; userId: string }
  | { kind: typeof SIGNER_KIND.external; email: string; name: string };

// NewSigningRequest is the create-request form payload.
export interface NewSigningRequest {
  document: File;
  signers: SignerSelection[];
  mode: SigningMode;
  recipientChannel: RecipientChannel;
  recipientAddress: string;
  recipientName: string;
  message: string;
}

function base(slug: string): string {
  return `/api/v1/orgs/${encodeURIComponent(slug)}/signing`;
}

export const signingAvailabilitySchema = z.object({ available: z.boolean() });
export type SigningAvailability = z.infer<typeof signingAvailabilitySchema>;

// getSigningAvailability reports whether a signing provider is configured for the
// org. It is member-safe (unlike the admin-only CSC settings), so it gates the
// signing nav item for every member who can actually use the feature.
export function getSigningAvailability(
  slug: string,
  signal?: AbortSignal,
): Promise<SigningAvailability> {
  return request(`${base(slug)}/availability`, {
    schema: signingAvailabilitySchema,
    signal,
  });
}

// getSigningCredential returns null when no credential is linked yet (the backend
// answers 404), so "not linked" is a normal state the UI renders, not an error.
export async function getSigningCredential(
  slug: string,
  signal?: AbortSignal,
): Promise<LinkedCredential | null> {
  try {
    return await request(`${base(slug)}/credential`, {
      schema: linkedCredentialSchema,
      signal,
    });
  } catch (error) {
    if (error instanceof ApiError && error.status === NOT_FOUND_STATUS) {
      return null;
    }
    throw error;
  }
}

export function linkSigningCredential(slug: string): Promise<SigningStart> {
  return request(`${base(slug)}/credential/link`, {
    schema: signingStartSchema,
    method: "POST",
  });
}

// createSigningRequest uploads a PDF plus the selected signers, mode and recipient,
// and returns the new request id. It does not sign — signers sign later. Each signer
// goes over as its own JSON `signers` value, so one list carries both kinds in the
// chosen order.
export function createSigningRequest(
  slug: string,
  input: NewSigningRequest,
): Promise<{ id: string }> {
  const form = new FormData();
  form.append("document", input.document);
  for (const signer of input.signers)
    form.append("signers", JSON.stringify(signer));
  form.append("mode", input.mode);
  form.append("recipientChannel", input.recipientChannel);
  form.append("recipientAddress", input.recipientAddress);
  form.append("recipientName", input.recipientName);
  form.append("message", input.message);
  return request(`${base(slug)}/requests`, {
    schema: createdSchema,
    method: "POST",
    body: form,
  });
}

// startSignRequest begins the acting user's signing ceremony for a request.
export function startSignRequest(
  slug: string,
  id: string,
): Promise<SigningStart> {
  return request(`${base(slug)}/requests/${encodeURIComponent(id)}/sign`, {
    schema: signingStartSchema,
    method: "POST",
  });
}

export function getSigningRequest(
  slug: string,
  id: string,
  signal?: AbortSignal,
): Promise<SigningRequest> {
  return request(`${base(slug)}/requests/${encodeURIComponent(id)}`, {
    schema: signingRequestSchema,
    signal,
  });
}

// getPendingSigningRequests lists the documents awaiting the caller's signature.
export async function getPendingSigningRequests(
  slug: string,
  signal?: AbortSignal,
): Promise<SigningRequest[]> {
  const page = await request(`${base(slug)}/requests/pending`, {
    schema: signingRequestListSchema,
    signal,
  });
  return page.requests ?? [];
}

// listSigningRequests returns the org-wide signing-request history, cursor-paginated.
export function listSigningRequests(
  slug: string,
  cursor?: string,
  signal?: AbortSignal,
): Promise<SigningRequestPage> {
  const query = cursor ? `?cursor=${encodeURIComponent(cursor)}` : "";
  return request(`${base(slug)}/requests${query}`, {
    schema: signingRequestPageSchema,
    signal,
  });
}

// The external-signee flow. An external signee has no membership and no session, so
// every call here is keyed by the one-time invitation token from their mail; the
// backend resolves it to a single signer row and takes nothing else from the caller.
function externalBase(token: string): string {
  return `/api/v1/signing/external/${encodeURIComponent(token)}`;
}

export const externalSigningSchema = z.object({
  orgName: z.string(),
  filename: z.string(),
  signerName: z.string(),
  signerEmail: z.string(),
  message: z.string().optional(),
  status: z.string(),
  signerStatus: z.string(),
  mode: z.string(),
  signerCount: z.number(),
  signedCount: z.number(),
  hasCredential: z.boolean(),
  canSign: z.boolean(),
  createdAt: z.string(),
});
export type ExternalSigning = z.infer<typeof externalSigningSchema>;

export function getExternalSigning(
  token: string,
  signal?: AbortSignal,
): Promise<ExternalSigning> {
  return request(externalBase(token), {
    schema: externalSigningSchema,
    signal,
  });
}

// linkExternalCredential starts the external signee's one-off credential-link
// ceremony (the same wallet ceremony a member runs from the "My credential" tab).
export function linkExternalCredential(token: string): Promise<SigningStart> {
  return request(`${externalBase(token)}/credential/link`, {
    schema: signingStartSchema,
    method: "POST",
  });
}

export function startExternalSign(token: string): Promise<SigningStart> {
  return request(`${externalBase(token)}/sign`, {
    schema: signingStartSchema,
    method: "POST",
  });
}

// externalDocumentUrl is the document the signee is being asked to sign, served
// inline so they can read it in the browser before signing.
export function externalDocumentUrl(token: string): string {
  return absoluteApiUrl(`${externalBase(token)}/document`);
}

// downloadSignedDocument fetches the signed PDF and triggers a browser save,
// using the server-suggested filename when present.
export async function downloadSignedDocument(
  slug: string,
  id: string,
  fallbackName: string,
): Promise<void> {
  const { blob, filename } = await requestBlob(
    `${base(slug)}/requests/${encodeURIComponent(id)}/document`,
  );
  const objectUrl = URL.createObjectURL(blob);
  try {
    const link = document.createElement("a");
    link.href = objectUrl;
    link.download = filename ?? fallbackName;
    document.body.appendChild(link);
    link.click();
    link.remove();
  } finally {
    URL.revokeObjectURL(objectUrl);
  }
}
